### ISSUE TITLE
zstd does #include within extern "C" scopes.  This breaks trying to build a clang module.

### ISSUE THREAD

**Issue by alecazam** [facebook/zstd#4187]
**Describe the bug**
#include should never be inside of scopes or namepsaces.  But zstd seems to extern "C" and then include various headers within them.  This then causes clang module building to fail.   I will trying moving all of these outside of these scopes, but in general this is bad practice for code used in so many applications..

import of C++ module 'std_stdint_h' appears within extern "C" language linkage specification

```
#if  !defined (__VMS) && (defined (__cplusplus) || (defined (__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L) /* C99 */) )
#  if defined(_AIX)
#    include <inttypes.h>
#  else
#    include <stdint.h> /* intptr_t */ <-
#  endif

Another example from the concatenated zstddeclib.cpp

#if defined (__cplusplus)
extern "C" {
#endif

/*-****************************************
*  Dependencies
******************************************/
#include <stddef.h>  /* size_t, ptrdiff_t */

```

**To Reproduce**
1. turn on -fmodules and -fcxx-modules 
2. setup a module.modulemap with zstd in it
3. Watch compile fail trying to build zstd

**Expected behavior**
All headers should just be included first and outside of any scoping.  Then any concatenation should just work.  extern "C" should only wrap function calls and structs for naming reasons.

**Screenshots and charts**
None

**Desktop (please complete the following information):**
 - OS: macOS 
 - Version 15.0
 - Compiler clang
 - Flags any
 - Other relevant hardware specs 8/4 cores
 - Build system Xcode

**Additional context**
None

**Comment by alecazam** [comment]
Alternatively, just have one file with all  #include and then prepend that at the top of the concatenated files.

**Comment by Cyan4973** [comment]
> example from the concatenated zstddeclib.cpp

Is the issue specific to the amalgamated single-file library ?

**Comment by alecazam** [comment]
Yes, from what I see one concatenated header will start an extern "C" scope, and then do #include of other zstd files.  So end up with #include nested inside of these scopes.   I've commented out the includes here.

This is from the top of zstd.h.   I've just commented out the includes.  But zstd.h, zstddeclib.cpp, and zstd.cpp all have similar patterns.

```
#if defined (__cplusplus)
extern "C" {
#endif

#ifndef ZSTD_H_235446
#define ZSTD_H_235446

/* ======   Dependency   ======*/
//#include <limits.h>   /* INT_MAX */
//#include <stddef.h>   /* size_t */
```

**Comment by Cyan4973** [comment]
The amalgamated single file libraries are supposed to be `*.c`.
Why are they renamed as `*.cpp`?

**Comment by Victor-C-Zhang** [comment]
Hey @alecazam, can you share the contents of your `module.modulemap` and the exact commands you used to repro the error?

**Comment by Victor-C-Zhang** [comment]
Update: I was unable to repro the build failure with the info provided.
Here's what I did:

1. Generate zstd.c from `create_single_file_library.sh`
2. Copy `zstd.c` and `lib/zstd.h`, `lib/zdict.h`, `lib/zstd_error.h`, `lib/module.modulemap` to a fresh directory (outside zstd repo) `~/module_test`
3. In the directory `~/module_test` use the following `main.cpp`:
```
#include "zzz.h"
#include <iostream>
#include <string>

int main() {
  std::cout << "Hello, world!" << std::endl;
  std::string input = "abcdabcdabcdabcdabcdabcdabcdabcd";
  std::string output('\0', input.size() * 100);
  auto res = ZSTD_compress(output.data(), output.size(),
                           input.data(), input.size(), 3);
  std::cout << res << std::endl;

  return 0;
}
```
4. For safety, rename the module `libzstd` -> `zzz`, and rename `zstd.h` -> `zzz.h`, `zstd.c` -> `zzz.c`. Also do this in the module.modulemap.
5. Build with module
```
clang -shared -o libzzz.dylib zzz.c
clang++ -std=c++20 main.cpp -fmodules -fcxx-modules -L. -lzzz
```
6. It runs:
```
% ./a.out
Hello, world!
18446744073709551546
```

Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin24.1.0
Thread model: posix

@alecazam can you share your repro steps?

**Comment by TroutZhang** [comment]
> Yes, from what I see one concatenated header will start an extern "C" scope, and then do #include of other zstd files. So end up with #include nested inside of these scopes. I've commented out the includes here.
> 
> This is from the top of zstd.h. I've just commented out the includes. But zstd.h, zstddeclib.cpp, and zstd.cpp all have similar patterns.
> 
> ```
> #if defined (__cplusplus)
> extern "C" {
> #endif
> 
> #ifndef ZSTD_H_235446
> #define ZSTD_H_235446
> 
> /* ======   Dependency   ======*/
> //#include <limits.h>   /* INT_MAX */
> //#include <stddef.h>   /* size_t */
> ```

Just met the same <limits.h> issue, due to enabling "-fmodules -fcxx-modules" today, in a cpp that has included "zstd.h"

**Comment by Victor-C-Zhang** [comment]
> Just met the same <limits.h> issue, due to enabling "-fmodules -fcxx-modules" today, in a cpp that has included "zstd.h"

Hi @TroutZhang, can you provide some brief repro steps as well as your Clang and machine info? I'm not sure how a `zstd.cpp` file was generated; this might be a quirk with some machines.

**Comment by alecazam** [comment]
Sorry, but I have all files in my open-source renamed to .cpp files.  That way I don't have to pull in C compilers, and only pull in cpp clang.   But I am pulling the zstd.h into other .cpp files.  But I think not having this extern "C" extend across includes would help with the C++ modules.  I haven't gotten all of this totally working with clang modules yet, and ran out of time to spend on it.  But will post more when I revisit it.

**Comment by TroutZhang** [comment]
> > Just met the same <limits.h> issue, due to enabling "-fmodules -fcxx-modules" today, in a cpp that has included "zstd.h"
> 
> Hi @TroutZhang, can you provide some brief repro steps as well as your Clang and machine info? I'm not sure how a `zstd.cpp` file was generated; this might be a quirk with some machines.

**clang and machine:**
Xcode 16.2 for clang

```
❯ clang --version 
Apple clang version 16.0.0 (clang-1600.0.26.6)
Target: arm64-apple-darwin24.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

machine is on macOS Sequoia 15.2.

Repro steps currently are hard to provide, as it's our internal codebase, I was just trying to enable the ""-fmodules -fcxx-modules" yesterday and found the issue, then googled here.


I currently workaround it by "-Wno-module-import-in-extern-c".

* Enabling "-fmodules"
<img width="327" alt="image" src="https://github.com/user-attachments/assets/35ebd8a3-f90e-49ae-ae7d-f6e0d496dca0" />

* Enabling "-fcxx-modules" as well as "-Wno-module-import-in-extern-c"
<img width="252" alt="image" src="https://github.com/user-attachments/assets/4066cd8b-34e1-43c1-a27d-a32b68fc3cc3" />
<img width="197" alt="image" src="https://github.com/user-attachments/assets/a4a48af9-fa04-4e76-acfa-ff9a7329211a" />

**Comment by TroutZhang** [comment]
Also @alecazam 

> All headers should just be included first and outside of any scoping.

I don't have much module experience. Is this some standard requirement for using modules in C++?

As zstd is not the only cause of our internal codebase build failures, I've also met "_-Wmodules-import-nested-redundant_" besides the "_-W-module-import-in-extern-c_" for zstd.

**Comment by Victor-C-Zhang** [comment]
> Repro steps currently are hard to provide, as it's our internal codebase, I was just trying to enable the ""-fmodules -fcxx-modules" yesterday and found the issue, then googled here.

@TroutZhang are you able to test if [this PR](https://github.com/facebook/zstd/pull/4218) fixes the problem? Otherwise, it would be helpful to get a minimum reproducible error, even a `Hello world` example with a simple compile command.

**Comment by TroutZhang** [comment]
@Victor-C-Zhang I understand you, I will first try testing the PR. If not working, I will try giving you a mini-repro.

It's just I am not currently familar with zstd and how it is used in our code base. I was doing a general module related test and found this issue. That's why there wasn't a repro in the first place.

**Comment by TroutZhang** [comment]
Hi @Victor-C-Zhang ,

For your PR, https://github.com/facebook/zstd/pull/4218/commits/fc726da7747b159520de2edb0febab30342d2744# 
Only the changes in **lib/zstd.h** is enough for me.

So, either of the following three, works in the codebase I worked with:
* your move-out of the inclusion to above  the "extern C"
* @alecazam 's comment-out of "#include <limits.h>"
* my work-around of disabling the warning by "-Wno-module-import-in-extern-c"

**Comment by Victor-C-Zhang** [comment]
Just merged #4218 into mainline. Based on @TroutZhang's testing, this should be a sufficient solution.

**Comment by Victor-C-Zhang** [comment]
Feel free to re-open if there are further issues.
